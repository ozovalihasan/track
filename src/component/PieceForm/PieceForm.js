import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const PieceForm = ({
  frequencyTimeList,
  frequencyTime,
  frequency,
  name,
  trackedItemList,
  trackedItemId,
  handleSubmit,
  handleChange,

}) => (
  <Form onSubmit={handleSubmit}>
    <input type="text" name="name" id="" value={name} onChange={handleChange} placeholder="Name" />
    <select name="trackedItemId" value={trackedItemId} onChange={handleChange}>
      {trackedItemList.map(trackedItem => (
        <option key={trackedItem.id} value={trackedItem.id}>
          {trackedItem.name}
        </option>
      ))}
    </select>
    <select name="frequencyTime" value={frequencyTime} onChange={handleChange}>
      {frequencyTimeList.map(time => (
        <option key={time[0]} value={time[1]}>
          {time[0]}
        </option>
      ))}
    </select>

    <input type="number" name="frequency" id="" value={frequency} onChange={handleChange} />
    <button type="submit"> Create Piece</button>
  </Form>

);

const Form = styled.form`
color: turquoise;
`;
PieceForm.propTypes = {
  frequencyTimeList: PropTypes.arrayOf().isRequired,
  frequencyTime: PropTypes.string.isRequired,
  frequency: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  handleChange: PropTypes.func.isRequired,
  trackedItemList: PropTypes.arrayOf(
    PropTypes.shape(
      {
        id: PropTypes.number,
        name: PropTypes.string,
      },
    ),
  ).isRequired,
  trackedItemId: PropTypes.string.isRequired,
};
export default PieceForm;
