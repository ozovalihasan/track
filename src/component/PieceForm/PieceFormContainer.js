import { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory } from 'react-router-dom';
import { fetchCreatePiece, fetchListTrackedItems } from '../../redux';
import PieceForm from './PieceForm';

const PieceFormContainer = () => {
  const frequencyTimeList = [['Day', '86400'], ['Week', '604800']];
  const [frequencyTime, setFrequencyTime] = useState(frequencyTimeList[0][1]);
  const [name, setName] = useState('');
  const [frequency, setFrequency] = useState('1');

  const trackedItemList = useSelector(state => state.trackedItem.list);
  const [trackedItemId, setTrackedItemId] = useState('');
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchListTrackedItems());
  }, []);

  useEffect(() => {
    if (trackedItemList.length !== 0) { setTrackedItemId(trackedItemList[0].id.toString()); }
  }, [trackedItemList]);

  const handleChange = e => {
    if (e.target.name === 'frequencyTime') {
      setFrequencyTime(e.target.value);
    } else if (e.target.name === 'name') {
      setName(e.target.value);
    } else if (e.target.name === 'frequency' && e.target.value >= 0) {
      setFrequency(e.target.value);
    } else if (e.target.name === 'trackedItemId') {
      setTrackedItemId(e.target.value);
    }
  };

  const history = useHistory();

  const handleSubmit = e => {
    e.preventDefault();
    dispatch(fetchCreatePiece({
      piece: {
        name, frequency, frequency_time: frequencyTime, percentage: 0,
      },
      tracked_item_id: trackedItemId,
    }));
    history.push('/');
  };
  return (
    <PieceForm
      frequencyTimeList={frequencyTimeList}
      frequencyTime={frequencyTime}
      frequency={frequency}
      name={name}
      trackedItemList={trackedItemList}
      trackedItemId={trackedItemId}
      handleSubmit={handleSubmit}
      handleChange={handleChange}

    />
  );
};

export default PieceFormContainer;
