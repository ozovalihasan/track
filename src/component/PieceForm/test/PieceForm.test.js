import { render, screen } from '@testing-library/react';
import React from 'react';
import '@testing-library/jest-dom';

import userEvent from '@testing-library/user-event';
import PieceForm from '../PieceForm';

let renderReadyComponent;
const frequencyTimeList = [['Day', '86400'], ['Week', '604800']];
const frequencyTime = '86400';
const frequency = '0';
const name = 'mock name';
const trackedItemList = [
  { id: 1, name: 'First Tracked Item' },
  { id: 2, name: 'Second Tracked Item' },
];
const trackedItemId = '';

let handleChange;
let handleSubmit;
let handleSubmitSpy;

beforeEach(() => {
  handleChange = jest.fn();
  handleSubmitSpy = jest.fn();
  handleSubmit = e => {
    handleSubmitSpy();
    e.preventDefault();
  };
  renderReadyComponent = (
    <PieceForm
      frequencyTimeList={frequencyTimeList}
      frequencyTime={frequencyTime}
      frequency={frequency}
      name={name}
      trackedItemList={trackedItemList}
      trackedItemId={trackedItemId}
      handleSubmit={handleSubmit}
      handleChange={handleChange}
    />
  );
});

afterEach(() => {
  handleChange.mockClear();
  handleSubmitSpy.mockClear();
});

describe('<PieceForm />', () => {
  it('call handleSubmit when the form is submitted', () => {
    render(renderReadyComponent);

    expect(handleSubmitSpy).not.toHaveBeenCalled();
    userEvent.click(screen.getByText(/Create Piece/i));
    expect(handleSubmitSpy).toHaveBeenCalled();
  });

  it('call handleChange when input of name is changed ', () => {
    render(renderReadyComponent);
    expect(handleChange).not.toHaveBeenCalled();
    userEvent.type(screen.getByDisplayValue(/mock name/i), '1');
    expect(handleChange).toHaveBeenCalled();
  });

  it('call handleChange when selected option of trackedItemId is changed ', () => {
    render(renderReadyComponent);
    expect(handleChange).not.toHaveBeenCalled();
    userEvent.selectOptions(screen.getByDisplayValue('Day'), ['Week']);
    expect(handleChange).toHaveBeenCalled();
  });

  it('call handleChange when selected option of frequencyTime is changed ', () => {
    render(renderReadyComponent);
    expect(handleChange).not.toHaveBeenCalled();
    userEvent.selectOptions(screen.getByDisplayValue('Day'), ['Week']);
    expect(handleChange).toHaveBeenCalled();
  });

  it('call handleChange when input of frequency is changed ', () => {
    render(renderReadyComponent);
    expect(handleChange).not.toHaveBeenCalled();
    userEvent.type(screen.getByDisplayValue(/0/i), '1');
    expect(handleChange).toHaveBeenCalled();
  });

  it('renders correctly', () => {
    const renderedContainer = render(renderReadyComponent);
    expect(renderedContainer).toMatchSnapshot();
  });
});
