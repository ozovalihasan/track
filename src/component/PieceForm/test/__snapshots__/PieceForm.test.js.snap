// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PieceForm /> renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="sc-bdfBwQ kBOUOb"
      >
        <input
          name="name"
          placeholder="Name"
          type="text"
          value="mock name"
        />
        <select
          name="trackedItemId"
        >
          <option
            value="1"
          >
            First Tracked Item
          </option>
          <option
            value="2"
          >
            Second Tracked Item
          </option>
        </select>
        <select
          name="frequencyTime"
        >
          <option
            value="86400"
          >
            Day
          </option>
          <option
            value="604800"
          >
            Week
          </option>
        </select>
        <input
          id=""
          name="frequency"
          type="number"
          value="0"
        />
        <button
          type="submit"
        >
           Create Piece
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="sc-bdfBwQ kBOUOb"
    >
      <input
        name="name"
        placeholder="Name"
        type="text"
        value="mock name"
      />
      <select
        name="trackedItemId"
      >
        <option
          value="1"
        >
          First Tracked Item
        </option>
        <option
          value="2"
        >
          Second Tracked Item
        </option>
      </select>
      <select
        name="frequencyTime"
      >
        <option
          value="86400"
        >
          Day
        </option>
        <option
          value="604800"
        >
          Week
        </option>
      </select>
      <input
        id=""
        name="frequency"
        type="number"
        value="0"
      />
      <button
        type="submit"
      >
         Create Piece
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
